proposals/KEEP-0389-kdoc-streamline-ambiguous-KDoc-links.md

```diff
@@ -0,0 +1,1752 @@
+
+Since some declarations might have overloads,
+KDoc names being able to point just to a single declaration feel too restricting.
+Sometimes it’s crucial to support this behavior
+(in the case with Dokka and other documentation-generating tools).
+However, we could enhance the user experience by showing a drop-down list with all the found overloads in the IDE.
+For use-cases when just a single declaration is required,
+users of the KDoc resolver could pick the first element from the resulting collection,
+which preserves the behavior.
+
+# Proposal
+
+This document proposes that existing KDoc links should point to only one
+(or multiple in case of overloads and some ambiguities) specific location in the same way 
+the Kotlin language does for name resolution in the code.
+
+Meanwhile, the ambiguity of KDoc links is still inevitable.
+Some might suggest providing an IDE warning informing users that their link is ambiguous, i.e.,
+that there are multiple declarations available for the given name.
+However, such an inspection would be useless and annoying, as there is no disambiguator at this time.
+Therefore, the proposed approach is to prioritize candidates for KDoc links 
+to deterministically choose a subset of all suitable declarations.
+
+To provide this disambiguation mechanism, the following questions should be covered:
+ - Which names are available from the current KDoc?
+ - If there is more than one candidate, what should be prioritized?
+ - If there are candidates that have the same priority, which should be chosen?
+
+The following sections address these three questions.
+
+## Context of KDoc
+
+The KDoc context is the set of all declarations that are resolvable at a specific, 
+imaginary place in the code of the documented property by the compiler.
+Hence, KDoc resolution rules derive from the language's resolution rules.
+
+For example, the KDoc context of a class is derived from the first line of a fake function body as such:
+```kotlin
+/**
+ * [p] is contained in the KDoc context
+ * [param] is not
+ */
+class A(param: Int) {
+    companion object {
+        val p = 0
+    }
+}
+```
+Here `[p]` is resolved since it is available on the first line of a fake function:
+```kotlin
+/**
+ * [p] is contained in the KDoc context
+ * [param] is not
+ */
+class A(param: Int) {
+    fun fake() {
+        p // is available here
+        param // ERROR: unresolved
+    }
+    
+    companion object {
+        val p = 0
+    }
+}
+```
+
+We can see that the constructor parameter `param` is not accessible in this context.
+It's not called "unresolved" as this link is a context declaration, so it's handled separately.
+
+Declarations don't have to be resolvable in the code as-is.
+It's contained in the KDoc context when there is a case this declaration is correctly referenced in the code.
+It might be a direct call, might be a reference, 
+but there must be a case when this declaration could be resolved in the code.
+```kotlin
```

Comment:
TBH, this example is awkward:
It could be rewritten:
```kt
class A {
    fun myFun() {}

    /**
     * [myFunFromC] is visible and resolved
     */
    class B {
        fun myFakeFun(x: C) {
            C.myFunFromC() // Valid
            C::myFunFromC // Valid
        }
    }
}

class C {
    fun myFunFromC() {}
}

```